### PR TITLE
PE-463: download folder

### PR DESCRIPTION
**Pull request by matibat**

### CODE REVIEW

File: src/ardrive_anonymous.ts [hunk 1]
```diff
@@ -80,11 +84,38 @@ export class ArDriveAnonymous extends ArDriveType {
...
 	}
 
-	async downloadPublicFile(publicFile: ArFSPublicFile, path: string): Promise<Stream> {
+	/**
+	 *
+	 * @param folderId - the ID of the folder to be download
+	 * @returns - the array of streams to write
+	 */
+	async downloadPublicFolder(folderId: FolderID, maxDepth: number, path: string): Promise<void> {
+		// const children = folderEntityDump.slice(1);
```
Comment (matibat) [inline src/ardrive_anonymous.ts:93]: remove unused code

File: src/folderHierarchy.ts [hunk 2]
```diff
@@ -1,6 +1,6 @@
 import { ArFSFileOrFolderEntity } from './arfs_entities';
 import { FolderID, EID } from './types';
-import { ROOT_FOLDER_ID_PLACEHOLDER } from './utils/arfs_builders/arfs_folder_builders';
+import { RootFolderID } from './utils/arfs_builders/arfs_folder_builders';
```
Comment (matibat) [inline src/folderHierarchy.ts:3]: **ToDo** revert these changes

**Review by matibat — COMMENTED** [review]
Please implement in this PR the tests for the stream class

**Review by matibat — COMMENTED** [review]
**For the reviewer**
Have in mind all the features implemented here for ArDrive and ArFSDAO are moved to https://github.com/ardriveapp/ardrive-core-js/pull/83

File: src/ardrive_anonymous.ts (hunk 1 above)
Comment (matibat) [reply src/ardrive_anonymous.ts:93]: ArDrive was de-coupled from CLI: https://github.com/ardriveapp/ardrive-core-js/pull/83

File: src/commands/download_folder.ts [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+} from '../parameter_declarations';
+
+new CLICommand({
+	name: 'download-folder',
+	parameters: [
+		FolderIdParameter,
+		LocalFilePathParameter,
+		DryRunParameter,
+		MaxDepthParameter,
+		{ name: DriveIdParameter, required: false },
```
Comment (arielmelendez) [inline src/commands/download_folder.ts:22]: Ooooh interesting.

File: src/commands/download_folder.ts [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+		...DrivePrivacyParameters
+	],
+	action: new CLIAction(async (options) => {
+		const parameters = new ParametersHelper(options);
+		const dryRun = !!parameters.getParameterValue(DryRunParameter);
+		const folderId = parameters.getRequiredParameterValue(FolderIdParameter, EID);
+		const localFilePath = parameters.getRequiredParameterValue(LocalFilePathParameter, resolve);
```
Comment (arielmelendez) [inline src/commands/download_folder.ts:29]: What features of `resolve` are you trying to lean on here?

File: src/commands/download_folder.ts [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+		const maxDepth = await parameters.getMaxDepth(Number.MAX_SAFE_INTEGER);
+
+		if (await parameters.getIsPrivate()) {
+			const driveId = parameters.getRequiredParameterValue(DriveIdParameter, EID);
+			const driveKey = await parameters.getDriveKey({ driveId });
+			const wallet = await parameters.getRequiredWallet();
+			const ardrive = cliArDriveFactory({
+				wallet,
+				feeMultiple: parameters.getOptionalBoostSetting(),
+				dryRun
```
Comment (arielmelendez) [inline src/commands/download_folder.ts:39]: How a downloading dry run work?

File: src/commands/download_folder.ts [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+			});
+			await ardrive.downloadPrivateFolder(folderId, maxDepth, localFilePath, driveKey);
+		} else {
+			const ardrive = cliArDriveAnonymousFactory({});
+			await ardrive.downloadPublicFolder(folderId, maxDepth, localFilePath);
```
Comment (arielmelendez) [inline src/commands/download_folder.ts:44]: If dry run is really a thing, why can't we do it with public folders?

File: src/utils/stream_decrypt.ts [hunk 7]
```diff
@@ -0,0 +1,39 @@
+import { CipherIV, FileKey } from 'ardrive-core-js';
+import { createDecipheriv } from 'crypto';
+import { Transform } from 'stream';
+
+const algo = 'aes-256-gcm'; // crypto library does not accept this in uppercase. So gotta keep using aes-256-gcm
+const authTagLength = 16;
+
+export class StreamDecrypt extends Transform {
```
Comment (arielmelendez) [inline src/utils/stream_decrypt.ts:8]: Should this still be here? Isn't it in the core PR now?

**Review by arielmelendez — CHANGES_REQUESTED** [review]
A few questions.

**Comment by matibat** [comment]
Outdated

**Pull request closed without merging**